### ISSUE TITLE
PRTE fails to link with hwloc 1.5.2 and 1.8.1

### ISSUE THREAD

**Issue by jsquyres** [open-mpi/ompi#9862]
[This was found in Cisco MTT](https://mtt.open-mpi.org/index.php?limit=&wrap=&trial=&enable_drilldowns=&yaxis_scale=&xaxis_scale=&hide_subtitle=&split_graphs=&remote_go=&do_cookies=&phase=mpi_install&text_start_timestamp=2022-01-10+16%3A30%3A01+-+2022-01-11+16%3A30%3A01&text_platform_hardware=%5Ex86_64%24&show_platform_hardware=show&text_os_name=%5ELinux%24&show_os_name=show&text_mpi_name=%5Eompi-nightly-master%24&show_mpi_name=show&text_mpi_version=%5Emaster-202201110241-2357d72%24&show_mpi_version=show&text_configure_arguments=&text_compiler_name=&show_compiler_name=show&text_bitness=&show_bitness=show&text_endian=&show_endian=show&text_http_username=%5Ecisco%24&show_http_username=show&text_local_username=all&show_local_username=hide&text_platform_name=%5Ecisco-community%2C+usNIC+build%3Dmtt-usnic-890.0%24&show_platform_name=show&click=Detail&phase=mpi_install&test_result=_mpi_f&text_compute_cluster_id=&text_os_version=&show_os_version=&text_platform_type=&show_platform_type=&text_submit_id=&text_hostname=&show_hostname=&text_mpi_get_id=&text_mpi_install_compiler_id=&text_mpi_install_configure_id=&text_mpi_install_id=&text_compiler_version=&show_compiler_version=&text_vpath_mode=&show_vpath_mode=&text_exit_value=&show_exit_value=&text_exit_signal=&show_exit_signal=&text_duration=&show_duration=&text_client_serial=&show_client_serial=&text_result_message=&text_result_stdout=&text_result_stderr=&text_environment=&text_description=&text_trial=%5Ef%24&lastgo=summary): building Open MPI master with an external hwloc 1.5.2 and 1.8.1

Here's the external hwloc 1.5.2 stderr:

```
$ ./configure --prefix=/home/jsquyres/bogus --with-hwloc=/home/mpiteam/hwloc-1.5.2/install
...
$ make -j 32
...
hwloc/hwloc_base_util.c: In function ‘prte_hwloc_base_filter_cpus’:
hwloc/hwloc_base_util.c:280:15: warning: implicit declaration of function ‘hwloc_get_numanode_obj_by_os_index’; did you mean ‘hwloc_get_pu_obj_by_os_index’? [-Wimplicit-function-declaration]
         obj = hwloc_get_numanode_obj_by_os_index(topo, w);
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
               hwloc_get_pu_obj_by_os_index
hwloc/hwloc_base_util.c:280:13: warning: assignment to ‘hwloc_obj_t’ {aka ‘struct hwloc_obj *’} from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
         obj = hwloc_get_numanode_obj_by_os_index(topo, w);
             ^
  CCLD     libprrte.la
make[4]: Leaving directory `/home/jsquyres/git/o3/3rd-party/prrte/src'
Making all in tools/prted
make[4]: Entering directory `/home/jsquyres/git/o3/3rd-party/prrte/src/tools/prted'
  CC       prted.o
  CCLD     prted
../../../src/.libs/libprrte.so: undefined reference to `hwloc_get_numanode_obj_by_os_index'
collect2: error: ld returned 1 exit status
make[4]: *** [prted] Error 1
```

Note that the declaration problems are just warnings; there's a linker error that is the cause for `make` to fail.

Using external hwloc 1.8.1, the compile warnings / link error is the same as it is for external hwloc 1.5.2.

If the minimum hwloc version is simply higher than 1.8.1 these days, just let me know and I'll update Cisco's MTT (as long as the v5.0.x RMs are ok with increasing the minimum hwloc version -- that's a different conversation).

It would be nice to fail earlier (e.g., in `configure`) if we're using an hwloc that is unsupported, though.

Per discussion on the 11 Jan 2022 webex, marked this as "blocker".

**Comment by bgoglin** [comment]
hwloc_get_numanode_obj_by_os_index() was added in 1.10. It's easy to backport with something like:
```
#if HWLOC_API_VERSION < 0x10a00
hwloc_obj_t hwloc_get_numanode_obj_by_os_index(hwloc_topology_t topology, unsigned os_index)
{
  hwloc_obj_t obj = NULL;
  while ((obj = hwloc_get_next_obj_by_type(topology, HWLOC_OBJ_NUMANODE, obj)) != NULL)
    if (obj->os_index == os_index)
      return obj;
  return NULL;
}
#endif
```

**Comment by bwbarrett** [comment]
@rhc54 I couldn't remember what we decided while we were talking about this issue on the OMPI call, so that was the first thing I was going to check.  There's a bug here regardless, because either 1) we should support hwloc 1.5.2 or 2) the configure test should abort with a message that says it can't find a new enough configure.  I'm guessing the code for (2) is broken, which is why I had Jeff assign the issue to me...

**Comment by rhc54** [comment]
Borked configure could be right - does OMPI want to support 1.5.2? I don't really care - like I said, this patch is trivial, though IIRC there are some other impacts (we compile, but may see odd things in terms of support).

Thx for taking a look at the configure question. I'll post a PR with the patch from Brice as a backup plan.

**Comment by awlauria** [comment]
I remember the line we drew in the sane for v5.0.x being hwloc v1.11.

**Comment by rhc54** [comment]
Done - someone should update the OMPI master submodule pointers to pick it up.

**Comment by jsquyres** [comment]
Ok.  I've updated Cisco's MTT for master / v5.0.x to only test with external hwloc 1.11.1.

Let's leave this issue open until the PRTE submodule pointer is updated here in OMPI.

**Comment by rhc54** [comment]
Should you update OMPI's configure to reject earlier versions of hwloc?

**Comment by bwbarrett** [comment]
yup; it's on my todo list.  Hence, leaving the ticket open.

**Comment by awlauria** [comment]
https://github.com/open-mpi/ompi/pull/9896 merged to v5.0.x.

Closing.
